File: DotNetGotchas/VBDN/CopyingObjects/Modified2/Copy/Person.vb
'Person.vb
Public Class Person
    Private theAge As Integer
    Private theBrain As Brain

    Public Sub New(ByVal age As Integer, ByVal aBrain As Brain)
        theAge = age
        theBrain = aBrain
    End Sub

    Public Sub New(ByVal another As Person)
        theAge = another.theAge

        If TypeOf another.theBrain Is SmarterBrain Then
            theBrain = New SmarterBrain( _
                CType(another.theBrain, SmarterBrain))
        Else
            theBrain = New Brain(another.theBrain)
        End If
    End Sub

    Public Overrides Function ToString() As String
        Return "This is person with age " & _
            theAge & " and " & _
            theBrain.ToString()
    End Function
End Class

File: DotNetGotchas/CSharp/CopyingObjects/Modified2/Copy/Person.cs
//Person.cs

using System;

namespace Copy
{
	public class Person
	{
		private int theAge;
		private Brain theBrain;

		public Person(int age, Brain aBrain)
		{
			theAge = age;
			theBrain = aBrain;
		}

		public Person(Person another)
		{
			theAge = another.theAge;

			if(another.theBrain is SmarterBrain)
			{
				theBrain = new SmarterBrain(
					(SmarterBrain) another.theBrain);
			}
			else
			{
				theBrain = new Brain(another.theBrain);
			}
		}

		public override string ToString()
		{
			return "This is person with age " +
					   theAge + " and " + 
                       theBrain;
		}

	}
}
